SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/ReferenceCountedLockingPolicy.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "ReferenceCountedLockingPolicy.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "Interop/SpatialStaticComponentView.h"
+#include "Schema/AuthorityIntent.h"
+
+#include "GameFramework/Actor.h"
+
+DEFINE_LOG_CATEGORY(LogReferenceCountedLockingPolicy);
+
+ActorLockToken UReferenceCountedLockingPolicy::AcquireLock(const AActor* Actor, FString DebugString)
+{
+	check(Actor != nullptr);
+	if (!CanAcquireLock(Actor)) {
+		UE_LOG(LogReferenceCountedLockingPolicy, Error, TEXT("Called AcquireLock when CanAcquireLock returned false. Actor: %s."), *Actor->GetName());
+		return SpatialConstants::INVALID_ACTOR_LOCK_TOKEN;
+	}
+	++ActorToReferenceCount.FindOrAdd(Actor);
+	UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Acquiring migration lock. "
+		"Actor: %s. Lock name: %s. Token %d: Locks held: %d."), *Actor->GetName(), *DebugString, NextToken, *ActorToReferenceCount.Find(Actor));
+	TokenToNameAndActor.Emplace(NextToken, LockNameAndActor{ MoveTemp(DebugString), Actor });
+	return NextToken++;
+}
+
+bool UReferenceCountedLockingPolicy::CanAcquireLock(const AActor* Actor) const
+{
+	check(Actor != nullptr);
+	const auto* NetDriver = Cast<USpatialNetDriver>(Actor->GetWorld()->GetNetDriver());
+	const auto EntityId = NetDriver->PackageMap->GetEntityIdFromObject(Actor);
+
+	const bool HasAuthIntent = NetDriver->VirtualWorkerTranslator->GetLocalVirtualWorkerId() ==
+			NetDriver->StaticComponentView->GetComponentData<SpatialGDK::AuthorityIntent>(EntityId)->VirtualWorkerId;
+	const bool HasAuth = NetDriver->StaticComponentView->GetAuthority(EntityId, SpatialConstants::AUTHORITY_INTENT_COMPONENT_ID) == WORKER_AUTHORITY_AUTHORITATIVE;
+	if (!HasAuth)
+	{
+		UE_LOG(LogReferenceCountedLockingPolicy, Log, TEXT("Can not lock actor migration. Do not have authority. Actor: %s"), *Actor->GetName());
```

Comment:
It seems useful to log who does have authority here. (also for intent below) This will also likely need to be Verbose long-term.

Reply:
It means this function has to have a permanent dependency on the static component view existing. What were you thinking this information would be useful for?

Reply:
Reading the comment about entities being created we might need that dependency anyway given the entity first being created case though.

Reply:
The dependency is annoying, so I won't push for this too hard. Mostly I'm thinking that we have the debug strings if something is locked, but those only work if you query the policy on the authoritative worker. This would tell you which worker was authoritative and you should look for locks on.

Reply:
I think that these would only be called if you want to check if you have "actual authority" before trying to lock, or if you have made a mistake and tried to lock when you weren't allowed to though. I'm not sure when you'd call this, it not just be that you need to fix your code, and it be useful to know which worker had authority.

Reply:
Taking the conclusion here as a no-op for trying to print the actual worker authoritative ID, will change logs to verbose.